File: packages/core/src/platform/events/bus.ts
import type { DomainEvent, DomainEventType } from "@agently/contracts";

/**
 * In-process event bus — the realtime backbone, deliberately abstracted behind
 * an interface so it can be swapped for Postgres LISTEN/NOTIFY, Supabase
 * Realtime, Redis pub/sub, or Kafka without touching producers or consumers.
 *
 * It keeps a bounded ring buffer of recent events so a reconnecting SSE/WebSocket
 * client can replay anything it missed since its `lastEventId` — the same
 * "resume from cursor" guarantee a durable broker gives, good enough for the
 * single-process dev/mock backend.
 */

export type EventHandler = (event: DomainEvent) => void;

export interface EventBus {
  publish(event: DomainEvent): void;
  /** Subscribe to all events. Returns an unsubscribe function. */
  subscribe(handler: EventHandler): () => void;
  /** Subscribe to a single event type with a narrowed handler. */
  on<T extends DomainEventType>(
    type: T,
    handler: (event: Extract<DomainEvent, { type: T }>) => void,
  ): () => void;
  /** Events buffered after the given event id (exclusive), oldest first. */
  replayAfter(eventId: string | null): DomainEvent[];
}

export interface EventBusOptions {
  /** Max events retained for replay. */
  bufferSize?: number;
}

export function createEventBus(opts: EventBusOptions = {}): EventBus {
  const bufferSize = opts.bufferSize ?? 500;
  const handlers = new Set<EventHandler>();
  const buffer: DomainEvent[] = [];

  return {
    publish(event) {
      buffer.push(event);
      if (buffer.length > bufferSize) buffer.shift();
      for (const h of handlers) {
        try {
          h(event);
        } catch {
          // A failing subscriber must never break the producer or its peers.
        }
      }
    },
    subscribe(handler) {
      handlers.add(handler);
      return () => handlers.delete(handler);
    },
    on(type, handler) {
      const wrapped: EventHandler = (e) => {
        if (e.type === type) handler(e as Extract<DomainEvent, { type: typeof type }>);
      };
      handlers.add(wrapped);
      return () => handlers.delete(wrapped);
    },
    replayAfter(eventId) {
      if (!eventId) return [...buffer];
      const idx = buffer.findIndex((e) => e.id === eventId);
      return idx === -1 ? [...buffer] : buffer.slice(idx + 1);
    },
  };
}

File: apps/api/internal/platform/bus.go
package platform

import (
	"sync"

	"github.com/agently/api/internal/domain"
)

// In-process event bus — the realtime backbone, abstracted behind an interface
// so it can be swapped for Postgres LISTEN/NOTIFY, Redis pub/sub, etc. without
// touching producers or consumers. Mirrors
// packages/core/src/platform/events/bus.ts.
//
// It keeps a bounded ring buffer of recent events so a reconnecting SSE client
// can replay anything it missed since its lastEventId.

type EventHandler func(event domain.DomainEvent)

type EventBus interface {
	Publish(event domain.DomainEvent)
	// Subscribe to all events; returns an unsubscribe function.
	Subscribe(handler EventHandler) func()
	// ReplayAfter returns events buffered after the given event id (exclusive),
	// oldest first. An empty id replays the whole buffer.
	ReplayAfter(eventID string) []domain.DomainEvent
}

const defaultBufferSize = 500

type memoryBus struct {
	mu         sync.RWMutex
	handlers   map[int]EventHandler
	nextID     int
	buffer     []domain.DomainEvent
	bufferSize int
}

// NewEventBus creates an in-memory event bus.
func NewEventBus() EventBus {
	return &memoryBus{
		handlers:   make(map[int]EventHandler),
		buffer:     make([]domain.DomainEvent, 0, defaultBufferSize),
		bufferSize: defaultBufferSize,
	}
}

func (b *memoryBus) Publish(event domain.DomainEvent) {
	b.mu.Lock()
	b.buffer = append(b.buffer, event)
	if len(b.buffer) > b.bufferSize {
		b.buffer = b.buffer[len(b.buffer)-b.bufferSize:]
	}
	// Snapshot handlers under the lock so we can invoke them after releasing it.
	handlers := make([]EventHandler, 0, len(b.handlers))
	for _, h := range b.handlers {
		handlers = append(handlers, h)
	}
	b.mu.Unlock()

	for _, h := range handlers {
		// A failing subscriber must never break the producer or its peers.
		func() {
			defer func() { _ = recover() }()
			h(event)
		}()
	}
}

func (b *memoryBus) Subscribe(handler EventHandler) func() {
	b.mu.Lock()
	id := b.nextID
	b.nextID++
	b.handlers[id] = handler
	b.mu.Unlock()

	return func() {
		b.mu.Lock()
		delete(b.handlers, id)
		b.mu.Unlock()
	}
}

func (b *memoryBus) ReplayAfter(eventID string) []domain.DomainEvent {
	b.mu.RLock()
	defer b.mu.RUnlock()
	if eventID == "" {
		out := make([]domain.DomainEvent, len(b.buffer))
		copy(out, b.buffer)
		return out
	}
	for i, e := range b.buffer {
		if string(e.EventID()) == eventID {
			out := make([]domain.DomainEvent, len(b.buffer)-i-1)
			copy(out, b.buffer[i+1:])
			return out
		}
	}
	// Unknown id — replay everything (same fallback as the TS bus).
	out := make([]domain.DomainEvent, len(b.buffer))
	copy(out, b.buffer)
	return out
}
